Css transition for expandable block

LostABike opened the pull request:
<!--
Thank you for your contribution to iTwinUI.

If you are only making changes to the docs site using the "Edit page on GitHub" link,
then you can ignore most of this template.
-->

## Changes

<!--
What kind of code changes does this PR include?
Mention anything that could be helpful for reviewers and include screenshots for visual changes.
-->

- Removed `WithCSSTransition` for `iui-expandable-content`
- Handle content transitioning in css code
- Removed all mixins

## Testing

<!--
How did you test your changes?
If your PR has visual changes, then make sure they are demonstrated in html test pages as well as
storybook, then approve visual test images for both (`yarn approve:css` and `yarn approve:react`).

If not applicable, you can write "N/A".
-->
- Manual testing on storybook
- Ran `yarn test` (approval TBD)

## Docs

<!--
If your PR includes user-facing changes, then update docs in all places (JSDoc, website, stories, etc).
Make sure to include a changeset (`yarn changeset`).

If not applicable, you can write "N/A".
-->
- Added changeset

mayank99 commented on packages/itwinui-css/src/expandable-block/expandable-block.scss:
```diff
@@ -195,6 +195,21 @@
   }
 
   .iui-expandable-content {
+    display: grid;
+  }
+
+  &[data-iui-expanded='false'] .iui-expandable-content {
+    grid-template-rows: 0fr;
+    overflow: hidden;
+    transition: grid-template-rows 0.2s, visibility 0.2s;
+    min-height: 0;
+    visibility: hidden;
+    @include iui-expandable-block-content;
+  }
+
+  &[data-iui-expanded='true'] .iui-expandable-content {
+    grid-template-rows: 1fr;
+    visibility: visible;
     @include iui-expandable-block-content;
   }
```
I've found a way to keep padding without needing two wrapping divs - it will require the use of `transform: translateY(-100%)` combined with `z-index` to hide the leftover padding under the header only when it's closed.

Also, because this requires changes in a few different elements, it would be best to clean up the CSS to remove the use of all mixins, and use classes directly. The mixins are confusing and unnecessary.

LostABike replied:
like..this? it works but not smooth. also removed mixin

mayank99 replied:
it's not smooth because:
(1) you have removed padding from the closed state.
(2) you are not transitioning `transform`.

this should work:
```scss
    display: grid;
    grid-template-rows: 0fr;
    overflow: hidden;
    z-index: -1;
    transition: grid-template-rows 0.2s, visibility 0.2s, transform 0.2s;

    > div {
      min-height: 0;
      visibility: hidden;
      transition: visibility 0.2s;
      padding: var(--_iui-expandable-block-content-padding);
    }
  }

  &[data-iui-expanded='true'] .iui-expandable-content {
    grid-template-rows: 1fr;

    > div {
      visibility: visible;
    }
  }

  &:not([data-iui-expanded='true']) .iui-expandable-content {
    transform: translateY(-100%);
  }
```

mayank99 replied:
as for the mixins, i really meant removing *all* mixins, not just `iui-expandable-content`.

LostABike replied:
Ok now the content is not visible altogether :-s I tried tweaking around too, setting visible variables and transformY on various divs/classes

mayank99 replied:
can you try to paste the exact code i wrote above? it works fine for me.

LostABike replied:
Yea using exact code caused the other issue of content now showing when expanded for me which is why I tweaked around (it is the one being pushed just now, you can check [here](https://itwin.github.io/iTwinUI/1396/react/?path=/story/core-expandableblock--basic)
![expandable block](https://github.com/iTwin/iTwinUI/assets/22462084/30cd3e0c-eb1e-449c-bc1a-7dd6ecdc3a4f)

I think is because z-index is still -1 from default. making it not -1 when `data-iui-expanded='true'` would show the content. But like I mentioned, animation is still not smooth cuz collapsing is instant:
![expandable block z](https://github.com/iTwin/iTwinUI/assets/22462084/bf8670da-807e-4565-8325-ccdab6cfbe8c)

mayank99 replied:
try adding `isolation: isolate` to `iui-expandable-block` (root)

LostABike replied:
sorcery?? thanks! that does it

mayank99 replied:
the problem was that `z-index: -1` places it underneath the background, so `isolation: isolate` will create a new stacking context which tells it to only place it underneath the elements within this context.

LostABike replied:
ohhh that makes sense!

mayank99 commented on apps/storybook/cypress-visual-screenshots/baseline/ExpandableBlock.test.ts-Accordion (Closed).png:
looks like it is occupying some extra space in collapsed state :(

![](https://github.com/iTwin/iTwinUI/assets/9084735/ae14b707-3022-4a10-9ff6-e38ea7bffc47)

LostABike replied:
ya I think it is because of the content's padding. I moved the padding to when expanded = true only. Also its bttom border was showing so I moved that as well. I think it looks correct now

mayank99 replied:
it's causing some pretty unpleasant content shift. try splitting the padding so that `padding-inline` is always applied, and `padding-block` is only applied when open.

<img width="442" alt="" src="https://github.com/iTwin/iTwinUI/assets/9084735/a9b3094b-12c5-4152-b6a7-1abf4b25b959">

LostABike replied:
Updated to your suggested change and test image. Seems ok now?

mayank99 commented on .changeset/itchy-poems-shout.md:
```diff
@@ -0,0 +1,6 @@
+---
+'@itwin/itwinui-react': minor
+'@itwin/itwinui-css': minor
+---
+
+Expandable block animation now uses pure css for transitioning and `WithCSSTransition` has been removed.
```
should be a patch change and should not mention internal names like "WithCSSTransition"

```suggestion
---
'@itwin/itwinui-react': patch
'@itwin/itwinui-css': patch
---

Expandable block animation now uses pure css instead of javascript for transitioning.
```

mayank99 commented on packages/itwinui-css/src/expandable-block/expandable-block.scss:
```diff
@@ -2,219 +2,195 @@
   border-radius: var(--iui-border-radius-1);
 
   &[data-iui-expanded='true'] {
     --_iui-expandable-block-expander-icon-transform: rotate(90deg);
```
this can be combined with the data-iui-expanded rule above (line 21)

LostABike replied:
Right, combined

mayank99 commented on packages/itwinui-css/src/expandable-block/expandable-block.scss:
```diff
@@ -2,219 +2,195 @@
+    display: grid;
+    grid-template-rows: 0fr;
+    overflow: hidden;
+    z-index: -1;
```
i think lets remove this and instead use `z-index: 1` on the expandable-block-header.

because the user might have some content that uses z-index and `-1` could potentially cause unexpected issues with it.

LostABike replied:
Moved to header

mayank99 commented on packages/itwinui-css/src/expandable-block/expandable-block.scss:
```diff
@@ -2,219 +2,195 @@
+    grid-template-rows: 0fr;
+    overflow: hidden;
+    z-index: -1;
+    transition: grid-template-rows 0.2s, visibility 0.2s, transform 0.2s;
```
lets use `--iui-duration-1` instead of `0.2s` everywhere in this file

also add some easing curve. `transition-timing-function: ease-out;` might be enough

mayank99 replied:
also we are missing a check for [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion), so that will need to be added to all transitions

LostABike replied:
Like this?

LostABike replied:
On Windows I can toggle animation (I think we did this before) and I think it works as intended. No transition animation when it is off.

mayank99 commented on packages/itwinui-css/src/expandable-block/expandable-block.scss:
```diff
@@ -2,219 +2,195 @@
+      visibility: hidden;
+      padding-inline: var(--_iui-expandable-block-content-padding);
+      transition: visibility 0.2s ease-out;
+      align-self: start;
```
this might be subjective but it looks slightly better with `end` (could be an illusion though)

```suggestion
      align-self: end;
```

mayank99 commented on packages/itwinui-css/src/expandable-block/expandable-block.scss:
```diff
@@ -2,219 +2,195 @@
-
-    background-color: var(--_iui-expandable-block-header-background-color-hover);
 
+  > .iui-expandable-header {
```
almost all of the classes in this file can be _unnested_ because they have descriptive unique names now

LostABike replied:
Removed nesting

mayank99 replied:
it's not removed though.

```scss
// original
.iui-expandable-block {
  // ...

  > .iui-expandable-header {
    // ...
  }
}
```

```scss
// expected
.iui-expandable-block {
  // ...
}
.iui-expandable-header {
  // ...
}
```

```scss
// what you did (currently)
.iui-expandable-block {
  // ...

  .iui-expandable-header {
    // ...
  }
}
```

LostABike replied:
hmm so I moved `iui-expandable-content` and `iui-expandable-header` out (haven't pushed), but for other things like `title` and `caption` under `label`, should we move those out too? I just mainly thought it looks more organized in a hierarchal way as is
![image](https://github.com/iTwin/iTwinUI/assets/22462084/1a346a09-9577-4c1f-a3e3-1fce8172a01d)

mayank99 replied:
> for other things like title and caption under label, should we move those out too?

yes, nesting should be avoided as much as possible because it leads to hard-to-maintain code

LostABike replied:
okay. I un-nested everything and tested functionality to make sure nothing broke :x

mayank99 replied:
for these two cases, you should be able to move the code back where it was and do nesting _opposite_ direction.
<img width="583" alt="image" src="https://github.com/iTwin/iTwinUI/assets/9084735/be29d060-05bf-469c-afe7-81e83c6e5ae7">

something like this:

```scss
.iui-expandable-content {
  // ...

  :where(.iui-expandable-block[data-iui-expanded='true']) & {
    // ...
  }

  :where(.iui-expandable-block:not([data-iui-expanded='true'])) & {
    // ...
  }
}
```

this ensures that all the code that is applied on `iui-expandable-content` is grouped together

LostABike replied:
oh cool! I was trying to achieve something similar a while back! thanks! updated

mayank99 commented on packages/itwinui-css/src/expandable-block/expandable-block.scss:
```diff
@@ -2,219 +2,195 @@
-    font-size: var(--iui-font-size-0);
-    color: var(--_iui-expandable-block-header-caption-color, var(--iui-color-text-muted));
+    > .iui-expandable-block-icon {
+      @include mixins.iui-icon-style('m');
```
**_unrelated_** to this PR. You should be able to remove this mixin and instead use `<Icon>` in react. This change can be done in #1354

LostABike replied:
as in wrap the element in `<Icon/>` instead of `<Box/>` and remove the css class?

mayank99 replied:
should be similar to `ExpandableBlockEndIcon`

LostABike replied:
okay I have updated this (might have forgotten to push yesterday) + resolved the merge conflicts. also re-tested to make sure things are still working as before

LostABike replied:
hold on css is messed up I'm looking into that

mayank99 reviewed (commented):
generally looks good so far! just a few minor comments

mayank99 commented on packages/itwinui-css/src/color-picker/color-picker.scss:
dont think these are intentional, should be reverted.

try running `yarn` so your dependencies are up-to-date, then save this file and push it.

LostABike replied:
ya weird, I reverted

mayank99 commented on packages/itwinui-css/src/toggle-switch/toggle-switch.scss:
these don't look intentional either.

LostABike replied:
reverted

LostABike commented:
@mayank99 added `flex-shrink: 0` to icon because it was getting squished within header container when title is too long. Also, positioning of icon was weird on html:
![image](https://github.com/iTwin/iTwinUI/assets/22462084/01184087-2161-4828-8b94-404baf424646)

so I removed `block-size: var(--_iui-expandable-block-expander-icon-size);`, which doesn't seem to do anything to other cases and work to fix small icon positioning

also fixed script within html, wasn't working correctly before.
